src/v/redpanda/request_auth.cc

```diff
@@ -0,0 +1,190 @@
+ * on destruction.
+ *
+ * @param req
+ * @return
+ */
+request_auth_result
+request_authenticator::authenticate(const ss::httpd::request& req) {
+    if (_controller == nullptr) {
+        // We are running outside of an environment with credentials, e.g.
+        // a unit test or a standalone pandaproxy/schema_registry
+        return request_auth_result(true, true);
+    }
+
+    const auto& cred_store = _controller->get_credential_store().local();
+    try {
+        return do_authenticate(req, cred_store, _require_auth());
+    } catch (ss::httpd::base_exception const& e) {
+        if (e.status() == ss::httpd::reply::status_type::unauthorized) {
+            if (_require_auth()) {
+                throw;
+            } else {
+                // Auth is disabled: give this user full access, but
+                // treat them as anonymous.
+                return request_auth_result(true, true);
+            }
+        } else {
+            throw;
+        }
+    }
+}
+
+request_auth_result request_authenticator::do_authenticate(
+  ss::httpd::request const& req,
+  security::credential_store const& cred_store,
+  bool require_auth) {
+    security::credential_user username;
+
+    auto auth_hdr = req.get_header("authorization");
+    if (auth_hdr.substr(0, 5) == "Basic") {
+        // Minimal length: Basic, a space, 1 or more bytes
+        if (auth_hdr.size() < 7) {
+            throw ss::httpd::bad_request_exception(
+              "Malformed Authorization header");
+        }
+
+        auto base64 = auth_hdr.substr(6);
+        ss::sstring decoded_bytes;
+        try {
+            decoded_bytes = base64_to_string(base64);
+        } catch (base64_decoder_exception const&) {
+            vlog(logger.info, "Client auth failure: bad BASE64 encoding");
+            throw ss::httpd::bad_request_exception(
+              "Malformed Authorization header");
+        }
+
+        auto colon = decoded_bytes.find(":");
+        if (colon == std::string::npos || colon == decoded_bytes.size() - 1) {
+            vlog(logger.info, "Client auth failure: malformed 'user:password'");
+            throw ss::httpd::bad_request_exception(
+              "Malformed Authorization header");
+        }
+        username = security::credential_user{decoded_bytes.substr(0, colon)};
+        auto password = ss::sstring(decoded_bytes.substr(colon + 1));
+
+        const auto cred_opt = cred_store.get<security::scram_credential>(
+          username);
+        if (!cred_opt.has_value()) {
+            for (auto i = cred_store.begin(); i != cred_store.end(); ++i) {
+                vlog(logger.info, "Credential: {}", i->first);
+            }
+            // User not found
+            vlog(
+              logger.info,
+              "Client auth failure: user '{}' not found",
+              username);
+            throw ss::httpd::base_exception(
+              "Unauthorized", ss::httpd::reply::status_type::unauthorized);
```

Comment:
I think the response must contain the header:
```
WWW-Authenticate: Basic
```

Reply:
You're right -- unfortunately the seastar httpd interface is a little bit limited here -- neither the exception handling path nor the json handler interface enables setting headers on the response.

I think the impact here is limited to things like web browsers rather than typical clients like RPK, so I think it's okay to defer this for right now: opened https://github.com/redpanda-data/redpanda/issues/3909